### PR TITLE
rgw: Improve error message on email id reuse

### PR DESCRIPTION
**Pull request by pponnuvel**
This corrects the patch in PR https://github.com/ceph/ceph/pull/39293

Fixes: https://tracker.ceph.com/issues/50554

The reproducer noted in PR https://github.com/ceph/ceph/pull/39293 is same here and has been verified with a local build.

Signed-off-by: Ponnuvel Palaniyappan <ponnuvel.palaniyappan@canonical.com>

Fixes: https://tracker.ceph.com/issues/50554

## Checklist
- [x] References tracker ticket
- [x] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by pponnuvel** [comment]
jenkins test make check

**Comment by pponnuvel** [comment]
@alimaredia @smanjara Can you review this please?

**Review by dang — APPROVED** [review]

**Review by cbodley — APPROVED** [review]

File: src/rgw/rgw_user.cc [hunk 1]
```diff
@@ -1876,7 +1876,14 @@ int RGWUser::add(const DoutPrefixProvider *dpp, RGWUserAdminOpState& op_state, o
 
   ret = check_op(op_state, &subprocess_msg);
   if (ret < 0) {
-    set_err_msg(err_msg, "unable to parse parameters, " + subprocess_msg);
+    if (is_populated() && (user_id.compare(op_state.get_user_id()) != 0)) {
+      set_err_msg(err_msg, "unable to create user " + user_id.to_str()
+		  + " because user id " + op_state.get_user_id().to_str()
+		  + " already exists with email "
+		  + op_state.get_user_email());
```
Comment (smanjara) [inline src/rgw/rgw_user.cc:1883]: by adding this condition here, we will end up printing this error message not only for duplicate email id, but also for other fields like duplicate `access_key` etc. 
The duplicate check already exists in `https://github.com/ceph/ceph/blob/master/src/rgw/rgw_user.cc#L1755` but it is masked by error returned from ` RGWUser::check_op() `

Comment (pponnuvel) [reply src/rgw/rgw_user.cc:1883]: Indeed. Perhaps it's better to have these checks in `RGWUser::check_op()` rather than in `RGWUser::execute_add()`.

**Comment by pponnuvel** [comment]
jenkins test api

**Comment by pponnuvel** [comment]
jenkins test api

**Review by dang — DISMISSED** [review]

**Comment by pponnuvel** [comment]
I am testing this myself and I'll confirm once that passes. Please don't merge until then.

**Comment by pponnuvel** [comment]
I am happy with my test. It now displays the error like:
```
 radosgw-admin user create --uid=test --display-name="AN test" --email=bar@domain.invalid
could not create user: unable to parse parameters, email: bar@domain.invalid is the email address of an existing user
```

If everything else is OK, I am happy to for this to be merged.

**Comment by pponnuvel** [comment]
jenkins test api

File: src/rgw/rgw_user.cc (hunk 1 above)
Comment (pponnuvel) [reply src/rgw/rgw_user.cc]: @smanjara ack, I've modified as such. Please take a look. Thanks.

**Comment by pponnuvel** [comment]
@ivancich It seems to be working after the latest changes. Please let me know if it breaks again.

**Comment by ivancich** [comment]
> @ivancich It seems to be working after the latest changes. Please let me know if it breaks again.

Out of curiosity, do you know why it broke? I removed my request for changes. When @smanjara gives the OK, we'll queue this up for QA again.

**Comment by pponnuvel** [comment]
@ivancich "radosgw-admin rm user ..." was failing because of the additional check mandating for display name (which is of course not needed for rm) in the previous commit. Taking those checks out of RGWUser::check_op() fixed it - now those checks are applied only for RGWUser::add().

**Review by dang — APPROVED** [review]

File: src/rgw/rgw_user.cc (hunk 1 above)
Comment (pponnuvel) [reply src/rgw/rgw_user.cc]: @smanjara ping

**Review by smanjara — APPROVED** [review]

**Review by ivancich — APPROVED** [review]

**Comment by ivancich** [comment]
@pponnuvel OK, I cleared my request. But it should go through QA again with the integration test changes.

**Comment by pponnuvel** [comment]
@yuriw Does this need any 'testing' labels? Just checking whether this PR is on track for QA.

**Comment by ivancich** [comment]
No regressions found in teuthology testing:
* https://pulpito.ceph.com/ivancich-2021-06-07_19:17:54-rgw-wip-eric-testing-1-distro-basic-gibba/
* https://pulpito.ceph.com/ivancich-2021-06-07_19:18:04-rgw-wip-eric-testing-1-distro-basic-gibba/

**Pull request merged**